app/Http/Controllers/ImportController.php

```diff
@@ -1,54 +1,80 @@
 <?php
 
-/** @noinspection PhpComposerExtensionStubsInspection */
-
-/** @noinspection PhpUndefinedClassInspection */
-
 namespace App\Http\Controllers;
 
 use App\Actions\Import\FromServer;
 use App\Actions\Import\FromUrl;
-use App\Http\Requests\ImportRequests\ImportServerRequest;
-use App\Http\Requests\ImportRequests\ImportUrlRequest;
+use App\Exceptions\MassImportException;
+use App\Http\Requests\Import\ImportFromUrlRequest;
+use App\Http\Requests\Import\ImportServerRequest;
+use App\Models\Photo;
+use Illuminate\Routing\Controller;
+use Illuminate\Support\Collection;
 use Illuminate\Support\Facades\Session;
-use ImagickException;
+use Symfony\Component\HttpFoundation\StreamedResponse;
 
+/**
+ * Class ImportController.
+ *
+ * This controller handles requests to import photos from a sequence of
+ * URLs or from a local path on the server.
+ *
+ * Note, the return types of the methods is quite differently.
+ * Both methods may take some time to finish, if the number of photos is large.
+ * While {@link ImportController::server()} returns a streamed response with
+ * intermediate updates, {@link ImportController::url} returns an atomic
+ * response at the end.
+ * The former approach appears to be more appropriate in order to avoid
+ * timeout errors at the client side.
+ * While {@link ImportController::url} returns a well-formed and
+ * syntactically valid JSON response which is a properly serialized collection
+ * of the imported photos, {@link ImportController::server()} just streams
+ * the terminal output of the command line.
+ * The latter cannot properly be interpreted by web-clients client which
+ * expect a proper JSON response.
+ * Actually, both methods should be re-factored such that their return
+ * types are consistent and the best of both worlds: a streamed collection.
+ * In other words, the streamed response should immediately send back a
+ * `[`-character to the client (the beginning of the collection), then
+ * send back a JSONized {@link Photo} as soon as it has been imported
+ * (element of the collection, and send a final `]`-character (the end of
+ * the collection).
```

Comment:
> Such format would effectively prevent the client from displaying the percentage of completion like it does now, although that percentage is not terribly useful as it is local to the current folder (i.e., for each subfolder it goes from 0 to 100).

No, it does not. It still works and even better than before. Just try it or have a look at the at the following code snippets: 

 - [`upload.start.server.action.runUpload.progressHandler`](https://github.com/LycheeOrg/Lychee-front/blob/d7f12fb3ccf94133c41111b47b14bea4c644e51e/scripts/main/upload.js#L658-L700)
 - [`upload.start.server.action.runUpload.processIncremental`](https://github.com/LycheeOrg/Lychee-front/blob/d7f12fb3ccf94133c41111b47b14bea4c644e51e/scripts/main/upload.js#L519-L526)

It is even much nicer and more type-safe than before, because we have proper JSON objects `ImportProgessReport` and `ImportEventReport` and use standard JSON encoding/decoding methods. So no problem with dangerous characters in strings (i.e. paths) anymore.
 
> More usefully, the current format shows which directory is being imported at the moment, and it provides meaningful errors/warnings ("Skipped duplicate XYZ", "Invalid file format XYZ", etc). Won't we lose that with what you propose?

Nope. See

- [`App\DTO\ImportReport`](https://github.com/LycheeOrg/Lychee/blob/6e13c20f644cf86a233ef617dc22928a8500e7bc/app/DTO/ImportReport.php)
- [`App\DTO\ImportProgressReport`](https://github.com/LycheeOrg/Lychee/blob/6e13c20f644cf86a233ef617dc22928a8500e7bc/app/DTO/ImportProgressReport.php)
- [`App\DTO\ImportEventReport`](https://github.com/LycheeOrg/Lychee/blob/6e13c20f644cf86a233ef617dc22928a8500e7bc/app/DTO/ImportEventReport.php)

for the backend side.

Basically, the backend streams a JSON array with proper report objects to the client. Report objects either indicate singular events or cumulative progress reports. The client receives this JSON array and progressively decodes the JSON stream (i.e. it decodes as much from the array as it can, because the array is truncated).

As a nice side effect we also get a readable response in the developer tools of the browser.

> And a bit of a historical perspective: I'm the one who added interim progress notifications on "Import from Server" to the web client. It was later extended to handle CLI as well (I believe we didn't even have CLI import capability at first). I never bothered with "Import from URL" probably because the interface does not encourage providing multiple URLs so I didn't think it was worth improving that aspect.

Not entirely correct. The interface does support multiple URLs, see

 - [`upload.start.url.action.runImport`](https://github.com/LycheeOrg/Lychee-front/blob/d7f12fb3ccf94133c41111b47b14bea4c644e51e/scripts/main/upload.js#L394-L421).
 - [`upload.start.dropbox.action.runImport`](https://github.com/LycheeOrg/Lychee-front/blob/d7f12fb3ccf94133c41111b47b14bea4c644e51e/scripts/main/upload.js#L835-L846)

It is only the frontend which just sends a single-valued array of URLs in case of import from ordinary URLs, because the dialog only allows to enter a single URL. But even for a single URL, I experienced timeout problems occasionally. I have never used the Dropbox feature myself, but this has to fail with a timeout for more than a coupe of files.

Reply:
Sorry, that's what I get for speaking up before looking over the whole PR... I misunderstood that you were planning to send back `Photo` objects only and nothing else...

I was actually under the impression that, at least with the front end on `master`, one _can_ provide multiple URLs by comma-separating them in the input box. But it's such a crude interface that I honestly can't imagine that anyone would bother, so I felt disinclined to improve server support for it.

As an aside, I am a little surprised how easily you seem to experience timeouts. I saw a mention of 500ms in one of the comments? I thought that just gives you a warning in the debug console -- don't true timeouts start at like 30s or even more?

Reply:
> Sorry, that's what I get for speaking up before looking over the whole PR... I misunderstood that you were planning to send back `Photo` objects only and nothing else...

No problem. Better be thoroughly. Besides adding a lot of jsDoc comments to the entire frontend code, most of the actual changes in the frontend went into revising the processing of the streamed response.

I am allergic to home-brewed encodings schemes, so I decided to replace it by something proper. It always tends to shoot you in the foot after some time (see the "XSS atack" reported recently, the problems with the markdown support and the unfortunate `lychee.html`-method).

I know that the whole code is "historically grown" or "legacy" and no single individual is responsible for the situation the code was in. I was also wondering if it was worth the effort to revise it with the emerging Livewire frontend at the horizon. But at least this approach also provides a template how the web installer can be improved and how the web installer may stream the actual console commands to the frontend.

> I was actually under the impression that, at least with the front end on `master`, one _can_ provide multiple URLs by comma-separating them in the input box. But it's such a crude interface that I honestly can't imagine that anyone would bother, so I felt disinclined to improve server support for it.

No at the moment you can't. However, there is a pending PR #1153 which tried to add such support for importing from the server. 

> As an aside, I am a little surprised how easily you seem to experience timeouts. I saw a mention of 500ms in one of the comments?

Good question. Maybe I set an option in "about:config" in the past which reduced the timeout value or it is system-dependent? I don't know. But I don't mind. It forces you to be careful what you code.

Reply:
No worries; I'm not emotionally attached to my home-brewed encoding scheme :wink: so I'm happy to see it replaced with something better.

FYI, I just tried entering two image URLs to "Import from Link" on `master` and it worked as I thought: both images were imported. It sounds like that's no longer the case with your PR? I'm sure we could fix that but honestly, I don't care: I wonder if more than three people in the world knew about that capability...